cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll

```diff
@@ -0,0 +1,590 @@
+    not exists(getValue(getConstantValue(x.getInstruction()))) and
+    if(strictlyPositive(x))
+    then (
+      upper = false and delta = 1
+    ) else
+      if positive(x)
+      then (
+        upper = false and delta = 0
+      ) else
+        if strictlyNegative(x)
+        then (
+          upper = true and delta = -1
+        ) else if negative(x) then (upper = true and delta = 0) else none()
+  )
+  or
+  exists(Operand x |
+    exists(SubInstruction sub |
+      i = sub and
+      sub.getAnOperand().(LeftOperand) = op and
+      sub.getAnOperand().(RightOperand) = x
+    )
+  |
+    // `x` with constant value is covered by valueFlowStep
+    not exists(getValue(getConstantValue(x.getInstruction()))) and
+    if strictlyPositive(x)
+    then (
+      upper = true and delta = -1
+    ) else
+      if positive(x)
+      then (
+        upper = true and delta = 0
+      ) else
+        if strictlyNegative(x)
+        then (
+          upper = false and delta = 1
+        ) else if negative(x) then (upper = false and delta = 0) else none()
+  )
+  or
+  i.(RemInstruction).getAnOperand().(RightOperand) = op and positive(op) and delta = -1 and upper = true
+  or
+  i.(RemInstruction).getAnOperand().(LeftOperand) = op and positive(op) and delta = 0 and upper = true
+  or
+  i.(BitAndInstruction).getAnOperand() = op and positive(op) and delta = 0 and upper = true
+  or
+  i.(BitOrInstruction).getAnOperand() = op and positiveInstruction(i) and delta = 0 and upper = false
+  // TODO: min, max, rand
+}
+
+private predicate boundFlowStepMul(Instruction i1, Operand op, int factor) {
+  exists(Instruction c, int k | k = getValue(getConstantValue(c)) and k > 0 |
+    i1.(MulInstruction).hasOperands(op, c.getAUse()) and factor = k
+    or
+    exists(ShiftLeftInstruction i |
+      i = i1 and i.getAnOperand().(LeftOperand) = op and i.getAnOperand().(RightOperand) = c.getAUse() and factor = 2.pow(k)
+    )
+  )
+}
+
+private predicate boundFlowStepDiv(Instruction i1, Operand op, int factor) {
+  exists(Instruction c, int k | k = getValue(getConstantValue(c)) and k > 0 |
+    exists(DivInstruction i |
+      i = i1 and i.getAnOperand().(LeftOperand) = op and i.getRightOperand() = c and factor = k
+    )
+    or
+    exists(ShiftRightInstruction i |
+      i = i1 and i.getAnOperand().(LeftOperand) = op and i.getRightOperand() = c and factor = 2.pow(k)
+    )
+  )
+}
+
+pragma[noinline]
+private predicate boundedNonPhiOperand(NonPhiOperand op, Bound b, int delta, boolean upper,
+  boolean fromBackEdge, int origdelta, Reason reason
+) {
+  exists(NonPhiOperand op2, int d1, int d2, Reason r1, Reason r2 |
+    boundFlowStepSsa(op, op2, d1, upper, r1) and
+    boundedNonPhiOperand(op2, b, d2, upper, fromBackEdge, origdelta, r2) and
+    delta = d1 + d2 and
+    (if r1 instanceof NoReason then reason = r2 else reason = r1)
+  )
+  or
+  boundedInstruction(op.getDefinitionInstruction(), b, delta, upper, fromBackEdge, origdelta, reason)
```

Comment:
This case here combined with the `op2.getDefinitionInstruction().getAnOperand().(CopySourceOperand) = op1` case in `boundFlowStepSsa` appears to duplicate some work.

Reply:
Fixed by removing the case in boundFlowStepSsa